src/main/java/net/blay09/mods/waystones/block/WaystoneBlock.java

```diff
@@ -91,48 +114,52 @@ public float getPlayerRelativeBlockHardness(BlockState state, PlayerEntity playe
         return super.getPlayerRelativeBlockHardness(state, player, world, pos);
     }
 
-    @Override
-    public boolean isValidPosition(BlockState state, IWorldReader world, BlockPos pos) {
-        // Do not allow placing a waystone directly on top of another
-        Block blockBelow = world.getBlockState(pos.down()).getBlock();
-        if (blockBelow == this) {
-            return false;
-        }
-
-        // Do not allow placing a waystone directly below of another
-        Block blockTwoAbove = world.getBlockState(pos.up(2)).getBlock();
-        BlockState stateAbove = world.getBlockState(pos.up());
-        return blockTwoAbove != this && stateAbove.isAir(world, pos.up());
-    }
-
     @Nullable
     @Override
     public BlockState getStateForPlacement(BlockItemUseContext context) {
         if (!PlayerWaystoneManager.mayPlaceWaystone(context.getPlayer())) {
             return null;
         }
 
-        return getDefaultState().with(FACING, context.getPlacementHorizontalFacing().getOpposite()).with(HALF, DoubleBlockHalf.LOWER);
+        final BlockPos pos = context.getPos();
+        final IFluidState ifluidstate = context.getWorld().getFluidState(context.getPos());
+
+        final BlockPos waystonePos = this.getWaystoneUpperPos(pos, context.getPlacementHorizontalFacing().getOpposite());
+        if (pos.getY() < 255 && waystonePos.getY() < 255 && context.getWorld().getBlockState(pos.up()).isReplaceable(
```

Comment:
Please use braces even for one-liner ifs, I find this really hard to read

Reply:
Good point